### ISSUE TITLE
[FEATURE] Support VXLAN network

### ISSUE THREAD

**Issue by guangbochen** [harvester/harvester#2322]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**
Add VXLAN network with adding the following advanced features:
- Support network policies for micro-segmentation of workloads #2260
- Support multiple VXLANs with separated physical NICs.
- Support VM live migration and persistent IP address.
- Integrate with kube-vip to provide load balancer service for VMs based on vxlan network. 
- Allow provisioning guest k8s cluster and manage the cluster with Rancher through the vxlan network.

**Comment by alejandroed** [comment]
It would be great to be able to use bgp unnumbered to do routing on the host and support full L3 fabrics, bonds are things of the past

**Comment by yasker** [comment]
Due to the large level of effort, we won't be able to make this into v1.1.0.

Move it to v1.2.0.

**Comment by moroniflores** [comment]
> Move it to v1.2.0.

When is v1.2.0 expected to be released?

**Comment by abonillabeeche** [comment]
> > Move it to v1.2.0.
> 
> When is v1.2.0 expected to be released?

Early Summer 2023 perhaps.

**Comment by yasker** [comment]
Yeah, H1 2023 is the planned date for v1.2.0

**Comment by yasker** [comment]
The VXLAN feature has been postponed to v1.3.0 which is targeting H2 2023.

**Comment by bk201** [comment]
Another request to ease the deployment on Hetzner https://github.com/harvester/harvester/issues/4622#issue-1945947885

> Hello together,
> I am currently playing with Harvester. My initial intent was to deploy Harvester in a production setting at a Hetzner datacenter. Unfortunately, I am seeing 2 show stopper for such a setup. Therefore, for now I switched to a Proxmox installation. Nevertheless, my hope is that soon missing pieces in Harvester come together, as it to my opinion is a very nice product. The first show stopper is described here https://github.com/harvester/harvester/issues/4608 , hence nothing to add.
> The second show stopper is related to the amount of MAC addresses I can use per switch port when using VLans (Hetzner vSwitch feature). The amount of MAC addresses is limited to 30 addresses per switch port. In other words, with one network card I can only support 30 VMs. In Proxmox I can simply create an overlay network with VxLan. Hence, this limitation is resolved with a little overhead. Would be great if VxLan could be made available in Harvester as well.

**Comment by mingshuoqiu** [comment]
Since we have the VLAN settings in the `VM Network` configuration, for the VxLAN setting, we may also need vlan-aware vxlan Network to passthrough vlan id inside the vxlan tunnel. 
Thus we can have 
1. Many vlans carried by one vxlan tunnel 
2. use vlan id mapping to vxlan vni (1-1 mapping for VNI and VLAN ID)

Does our VM Network configuration should includes both above?

**Comment by bk201** [comment]
cc @w13915984028 @starbops any feedback to Chris' comment is welcome.

**Comment by starbops** [comment]
My envision for this requirement is to utilize VLAN and VxLAN as providers for the Harvester VM Network. VM Network should be an abstration to the implementation details of the underlying network technolgies. The concept is similar to the mechanism driver under OpenStack Neutron, which supports various implementations such as flat, VLAN, VxLAN, GRE, etc. As we already have VLAN implementation (refer to VlanConfig CRD), we can design new CRDs and implement some logics against them to complement the corresponding part in the VxLAN context to address this aspect. Therefore, I believe we can initially keep things simple without considering the "VLAN in VxLAN" or vice versa scenarios. If there's a need for VLAN-based VM Networks to communicate with VxLAN-based VM Networks, it can be achieved through external routing. Just my two cents.

**Comment by w13915984028** [comment]
Thanks for the ideas.

The general VxLan is a big concept. Harvester will implement/utilize a sub-set of it to focus on Harvester's scenario.

Let's define the user story, alternative architecture in Harvester first, then it will come to the detailed design of CRDs/controllers... .

**Comment by mingshuoqiu** [comment]
Agreed. Let's start from the user story. From my understanding, the VMs of different Nodes should be able to connect to each other via `flanneld` by VxLAN when no VLAN configured. With VLAN configured, the VMs of different Nodes can reach each other via real world's switch/router in the same L2/L3 layer network. 
What do we want to achieve with the VxLAN enabled `VM Network`? Since VxLAN is L3 routable, we just need to determine the VTEP on different Nodes and make sure each VTEP is aware of how to reach each other. 
With VxLAN feature, how the user setup the Network?
1. Create a VM Network w/ VLAN ID (or untagged)
2. Create a local-router to play as a VTEP and a DHCP server
3. Render a configuration UI to identify the subnet of the specified VNI
4. The created local-router of each VM Network should be aware of each other by BGP-EVPN automatically? Or we should provide UI for how to connect VTEPs?
5. Is there any requirement of the connection between VMs from different Nodes?
Please suggest if there's anything need to be taken into consideration. Thanks

**Comment by mingshuoqiu** [comment]
Some follow-ups will be taken after discussion with @abonillabeeche 
1. Check the support status of Geneve under Linux
2. Create `vxlan` interfaces on different nodes resides at different subnet (or VLANs) to check the connectivity
3. Make sure VM with `vxlan` interface can http(s) the management url w/o problem. (since the http(s) are tunneled by the flannel interface).
4. Create `vxlan` interfaces on different Linux host and use VPN endpoint as VTEP to check the connectivity. (experimental)

I expect to have more Nodes resides at different network environments to work in the same harvester cluster. Please correct me if I misunderstand anything about customer's requirement.

**Comment by w13915984028** [comment]
@mingshuoqiu Please also take a look of the original HEP document. We will define the user stories and feature specifications. thanks.

https://github.com/w13915984028/harvester/blob/vxlan-hep/enhancements/20220530-vxlan-enhancement.md

**Comment by innobead** [comment]
Duplicate of https://github.com/harvester/harvester/issues/7364 cc @harvester/network
